File: src/main/webapp/app/programming/shared/services/legacy-build-plan-converter.service.ts
import { Injectable } from '@angular/core';
import { BuildPhase, BuildPlanPhases } from 'app/programming/shared/entities/build-plan-phases.model';
import { parseJson } from 'app/foundation/util/json.util';

/**
 * The purpose of this service is to transform the legacy build plan into the new format so it can still be edited
 * with the new editor.
 */
@Injectable({ providedIn: 'root' })
export class LegacyBuildPlanConverterService {
    private static readonly LOCAL_CI_DOCKER_CONTAINER_WORKING_DIRECTORY = '/var/tmp';

    convertLegacyBuildPlanConfiguration(script: string | undefined, buildPlanConfiguration: string | undefined): BuildPlanPhases | undefined {
        let parsed: Record<string, unknown> | undefined;
        const buildScript = script?.trim() ? script : undefined;

        if (buildPlanConfiguration?.trim()) {
            try {
                const parsedJson = parseJson(buildPlanConfiguration);
                if (this.isObject(parsedJson)) {
                    parsed = parsedJson;
                }
            } catch {
                if (buildScript === undefined) {
                    return undefined;
                }
            }
        }

        const actionsNode = parsed?.['actions'];
        if (buildScript === undefined && !Array.isArray(actionsNode)) {
            return undefined;
        }

        const convertedBuildScript = buildScript ?? this.createLegacyBuildScriptFromActions(actionsNode);

        return { phases: this.wrapLegacyBuildScript(convertedBuildScript, this.parseResultPaths(actionsNode)), dockerImage: parsed ? this.parseDockerImage(parsed) : undefined };
    }

    private isObject(value: unknown): value is Record<string, unknown> {
        return typeof value === 'object' && value !== null;
    }

    private parseDockerImage(parsed: Record<string, unknown>): string | undefined {
        const metadata = parsed['metadata'];
        if (!this.isObject(metadata)) {
            return undefined;
        }

        const docker = metadata['docker'];
        if (!this.isObject(docker)) {
            return undefined;
        }

        const dockerImageNode = docker['image'];
        if (typeof dockerImageNode !== 'string') {
            return undefined;
        }

        return dockerImageNode.trim();
    }

    private createLegacyBuildScriptFromActions(actionsNode: unknown): string {
        let buildScript = `#!/bin/bash
cd ${LegacyBuildPlanConverterService.LOCAL_CI_DOCKER_CONTAINER_WORKING_DIRECTORY}/testing-dir
`;
        if (!Array.isArray(actionsNode)) {
            return buildScript;
        }

        for (const actionNode of actionsNode) {
            if (!this.isObject(actionNode) || typeof actionNode['script'] !== 'string') {
                continue;
            }

            const workdir = typeof actionNode['workdir'] === 'string' ? actionNode['workdir'].trim() : undefined;
            if (workdir) {
                buildScript += `cd ${LegacyBuildPlanConverterService.LOCAL_CI_DOCKER_CONTAINER_WORKING_DIRECTORY}/testing-dir/${workdir}
`;
            }
            buildScript += `${actionNode['script']}
`;
            if (workdir) {
                buildScript += `cd ${LegacyBuildPlanConverterService.LOCAL_CI_DOCKER_CONTAINER_WORKING_DIRECTORY}/testing-dir
`;
            }
        }

        return buildScript;
    }

    private parseResultPaths(actionsNode: unknown): string[] {
        if (!Array.isArray(actionsNode)) {
            return [];
        }

        const resultPaths: string[] = [];
        for (const actionNode of actionsNode) {
            if (!this.isObject(actionNode)) {
                continue;
            }

            const resultsNode = actionNode['results'];
            if (resultsNode === undefined || resultsNode === null) {
                continue;
            }
            if (!Array.isArray(resultsNode)) {
                continue;
            }

            for (const resultNode of resultsNode) {
                if (!this.isObject(resultNode) || typeof resultNode['path'] !== 'string') {
                    continue;
                }

                resultPaths.push(resultNode['path'].trim());
            }
        }

        return resultPaths;
    }

    private wrapLegacyBuildScript(script: string, resultPaths: string[]): BuildPhase[] {
        const wrappedScript = `# feel free to remove the code surrounding your script and split your script into multiple phases
cd ${LegacyBuildPlanConverterService.LOCAL_CI_DOCKER_CONTAINER_WORKING_DIRECTORY}/testing-dir
local tmp_file=$(mktemp)
cat << '  __LEGACY_INNER_SCRIPT_END__' > "\${tmp_file}"  # two leading spaces are intentional as the final script will be indented be for a phase
${script}
__LEGACY_INNER_SCRIPT_END__
chmod +x "\${tmp_file}"
"\${tmp_file}" "$@"
`;
        return [
            {
                name: 'script',
                script: wrappedScript,
                condition: 'ALWAYS',
                forceRun: false,
                resultPaths,
            },
        ];
    }
}

File: src/main/java/de/tum/cit/aet/artemis/localci/service/LegacyBuildPlanConverterService.java
package de.tum.cit.aet.artemis.localci.service;

import static de.tum.cit.aet.artemis.core.config.Constants.LOCAL_CI_DOCKER_CONTAINER_WORKING_DIRECTORY;
import static de.tum.cit.aet.artemis.core.config.Constants.PROFILE_LOCALCI;

import java.util.ArrayList;
import java.util.List;
import java.util.Optional;

import org.springframework.context.annotation.Lazy;
import org.springframework.context.annotation.Profile;
import org.springframework.stereotype.Service;

import com.fasterxml.jackson.core.JsonProcessingException;
import com.fasterxml.jackson.databind.JsonNode;
import com.fasterxml.jackson.databind.ObjectMapper;

import de.tum.cit.aet.artemis.core.util.JsonObjectMapper;
import de.tum.cit.aet.artemis.programming.domain.ProgrammingExercise;
import de.tum.cit.aet.artemis.programming.domain.ProgrammingExerciseBuildConfig;
import de.tum.cit.aet.artemis.programming.domain.build.BuildPhaseCondition;
import de.tum.cit.aet.artemis.programming.dto.BuildPhaseDTO;
import de.tum.cit.aet.artemis.programming.dto.BuildPlanPhasesDTO;

/**
 * The purpose of this service is to transform the legacy build plan into the new phases format.
 */
@Lazy
@Service
@Profile(PROFILE_LOCALCI)
public class LegacyBuildPlanConverterService {

    private static final ObjectMapper objectMapper = JsonObjectMapper.get();

    /**
     * If successful it returns a present {@link BuildPlanPhasesDTO} containing the legacy build script wrapped into one build phase.
     *
     * @param programmingExercise the exercise that is assumed to be legacy
     * @return the converted build plan phases
     */
    public Optional<BuildPlanPhasesDTO> convertLegacyBuildPlanConfiguration(ProgrammingExercise programmingExercise) {
        return convertLegacyBuildPlanConfiguration(programmingExercise.getBuildConfig());
    }

    /**
     * If successful it returns a present {@link BuildPlanPhasesDTO} containing the legacy build script wrapped into one build phase. If the build script is missing,
     * the old Windfile script actions are converted into the legacy script first.
     *
     * @param buildConfig the build config that is assumed to be legacy
     * @return the converted build plan phases
     */
    public Optional<BuildPlanPhasesDTO> convertLegacyBuildPlanConfiguration(ProgrammingExerciseBuildConfig buildConfig) {
        if (buildConfig == null) {
            return Optional.empty();
        }

        String buildPlanConfiguration = buildConfig.getBuildPlanConfiguration();
        String buildScript = buildConfig.getBuildScript();
        JsonNode node = null;

        if (buildPlanConfiguration != null && !buildPlanConfiguration.isBlank()) {
            try {
                node = objectMapper.readTree(buildPlanConfiguration);
                if (!node.isObject()) {
                    node = null;
                }
            }
            catch (JsonProcessingException e) {
                if (buildScript == null) {
                    return Optional.empty();
                }
            }
        }

        JsonNode actionsNode = node == null ? null : node.path("actions");
        if (buildScript == null && !isLegacyWindfile(actionsNode)) {
            return Optional.empty();
        }

        if (buildScript == null) {
            buildScript = createLegacyBuildScriptFromActions(actionsNode);
        }

        String dockerImage = node == null ? null : parseDockerImage(node);
        List<String> resultPaths = parseResultPaths(actionsNode);

        return Optional.of(new BuildPlanPhasesDTO(wrapLegacyBuildScript(buildScript, resultPaths), dockerImage));
    }

    private static List<BuildPhaseDTO> wrapLegacyBuildScript(String script, List<String> resultPaths) {
        String wrappedScript = """
                # feel free to remove the code surrounding your script and split your script into multiple phases
                cd %s/testing-dir
                local tmp_file=$(mktemp)
                cat << '  __LEGACY_INNER_SCRIPT_END__' > "${tmp_file}"  # two leading spaces are intentional as the final script will be indented be for a phase
                %s
                __LEGACY_INNER_SCRIPT_END__
                chmod +x "${tmp_file}"
                "${tmp_file}" "$@"
                """.formatted(LOCAL_CI_DOCKER_CONTAINER_WORKING_DIRECTORY, script);
        return List.of(new BuildPhaseDTO("script", wrappedScript, BuildPhaseCondition.ALWAYS, false, resultPaths));
    }

    private static String parseDockerImage(JsonNode node) {
        JsonNode imageNode = node.path("metadata").path("docker").path("image");
        if (imageNode.isMissingNode() || imageNode.isNull()) {
            return null;
        }
        if (!imageNode.isTextual()) {
            return null;
        }
        return imageNode.asText().trim();
    }

    private static boolean isLegacyWindfile(JsonNode actionsNode) {
        return actionsNode != null && actionsNode.isArray();
    }

    private static String createLegacyBuildScriptFromActions(JsonNode actionsNode) {
        StringBuilder buildScriptBuilder = new StringBuilder();
        buildScriptBuilder.append("#!/bin/bash\n");
        buildScriptBuilder.append("cd ").append(LOCAL_CI_DOCKER_CONTAINER_WORKING_DIRECTORY).append("/testing-dir\n");

        if (!isLegacyWindfile(actionsNode)) {
            return buildScriptBuilder.toString();
        }

        for (JsonNode actionNode : actionsNode) {
            if (!actionNode.isObject() || !actionNode.path("script").isTextual()) {
                continue;
            }

            JsonNode workdirNode = actionNode.path("workdir");
            String workdir = workdirNode.isTextual() ? workdirNode.asText().trim() : null;
            if (workdir != null && !workdir.isBlank()) {
                buildScriptBuilder.append("cd ").append(LOCAL_CI_DOCKER_CONTAINER_WORKING_DIRECTORY).append("/testing-dir/").append(workdir).append("\n");
            }
            buildScriptBuilder.append(actionNode.path("script").asText()).append("\n");
            if (workdir != null && !workdir.isBlank()) {
                buildScriptBuilder.append("cd ").append(LOCAL_CI_DOCKER_CONTAINER_WORKING_DIRECTORY).append("/testing-dir\n");
            }
        }

        return buildScriptBuilder.toString();
    }

    private static List<String> parseResultPaths(JsonNode actionsNode) {
        if (!isLegacyWindfile(actionsNode)) {
            return List.of();
        }

        List<String> resultPaths = new ArrayList<>();
        for (JsonNode actionNode : actionsNode) {
            if (!actionNode.isObject()) {
                continue;
            }

            JsonNode resultsNode = actionNode.path("results");
            if (resultsNode.isMissingNode() || resultsNode.isNull()) {
                continue;
            }
            if (!resultsNode.isArray()) {
                continue;
            }

            for (JsonNode resultNode : resultsNode) {
                if (!resultNode.isObject()) {
                    continue;
                }
                JsonNode pathNode = resultNode.path("path");
                if (!pathNode.isTextual()) {
                    continue;
                }
                resultPaths.add(pathNode.asText().trim());
            }
        }

        return resultPaths;
    }
}
